keps/sig-node/753-keystone-containers/README.md

```diff
@@ -0,0 +1,815 @@
+        `alpha.node.kubernetes.io/keystone:{list of keystone containers}`
+2. One annotation per container
+   e.g. `alpha.sidecars.kubernetes.io/<container name>:string`
+
+The motivation for one annotation per container is easy translation to api fields later on and it can be used for the sidecar ordering proposals.
+
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+If we choose to specify sidecars container in pod annotation then the user can add all containers in a pod as sidecars and kubelet will terminate the pod. This can be thought of as intended behavior and misuse of annotation than risk. In future, a validation can be added for it.
+
+If we choose to specify keystone containers in pod annotation then even if a user specifies all containers as keystone, the pod will be keep running. And if the user has not specified any annotation kubelet will treat all containers as same as before.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We will add the logic to terminate pod on the basis of application container status, restart policy and exit code of application containers in `computePodActions` of kuberuntime manager. `computePodAction` is called by pod worker during pod sync. 
+
+
+```
+pkg/kubelet/kuberuntime/kuberuntime_manager.go 
+
+func (...) computePodActions(...){
+....
+....
+    if featureflag keystoneContainersEnabled 
+        keyContainers := getKeyContainersList(pod.Annotations["alpha.kubernetes.io/keycontainers"])
+        keyContainersStatus := getKeyContainersStatus(keyContainers)
+
+        if keyContainerStatus==Completed && restartPolicy != always
+        // kill other containers
```

Comment:
What if there are multiple keystone containers? What is the logic then? It's possible that there could be multiple, so I'm not sure how we collapsed their status down into a single state.

Reply:
In alpha we limit the number of keystone containers to 1, and will implement a validation during pod admission.

Reply:
This was not clear to me, please make sure this is stated explicitly, maybe in graduation criteria too.

Reply:
Yes, it's the case, sorry for the inconvenience... still waiting for the repository permissions.

Reply:
made the dummy code more clear, replaced `KeyContainers` with `KeyContainer` & @matthyx clarified in beta criteria as well, depending on the feedback we will remove limit for keystone containers